src/state.py

```diff
@@ -0,0 +1,107 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Jenkins States."""
+import logging
+import typing
+
+import ops
+
+logger = logging.getLogger(__name__)
+
+
+class ValidationError(Exception):
+    """An unexpected data is encountered."""
+
+
+class AgentMeta(typing.NamedTuple):
+    """Metadata for registering Jenkins Agent.
+
+    Attrs:
+        executors: Number of executors of the agent in string format.
+        labels: Comma separated list of labels to be assigned to the agent.
+        slavehost: The host name of the agent.
+    """
+
+    executors: str
+    labels: str
+    slavehost: str
+
+    def validate(self) -> None:
+        """Validate the agent metadata.
+
+        Raises:
+            ValidationError: if the field contains invalid data.
+        """
+        # Pylint doesn't understand that _fields is implemented in NamedTuple.
+        empty_fields = [
+            field
+            for field in self._fields  # pylint: disable=no-member
+            if not getattr(self, field)
+        ]
+        if empty_fields:
+            raise ValidationError(f"Fields {empty_fields} cannot be empty.")
+        try:
+            int(self.executors)
+        except ValueError as exc:
+            raise ValidationError(
+                f"Number of executors {self.executors} cannot be converted to type int."
+            ) from exc
+
+
+class State:
```

Comment:
Would making this class a `@dataclass(frozen=True)` simplify? It would avoid the `__init__` and properties, and make it easier to document each attribute, in just a few lines of code. You'd still need `from_charm`, of course.

Reply:
That did simplify the whole thing a lot! Fixed, thank you!